File: tune/volume/volume_test.c
/*
 * Copyright (c) 2017, Intel Corporation
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *   * Neither the name of the Intel Corporation nor the
 *     names of its contributors may be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
 */

#include "common_test.h"

/* Internal buffer */
#define TEST_BENCH_DEADLINE 1000 /* 1 ms */
#define TEST_BENCH_NCH 2 /* Stereo */
#define TEST_BENCH_RATE_MAX 48000 /* Hz */
#define TEST_BENCH_PERIOD_FRAMES (TEST_BENCH_RATE_MAX*TEST_BENCH_DEADLINE/1000000)
#define TEST_BENCH_FRAME_SIZE (4 * TEST_BENCH_NCH) /* 4bytes x  NCH */
#define INT_PERIOD_SIZE (TEST_BENCH_PERIOD_FRAMES * TEST_BENCH_FRAME_SIZE)

/* main firmware context */
static struct reef reef;

/*
 * Test Pipeline (ID=5)
 *
 * Fileread(0) ---B1(3)---> Volume(1) ---B2(4)---> Filewrite(2)
 */

#define FR_ID 0
#define VOLUME_ID 1
#define FW_ID 2
#define PIPE_ID 5
#define SCHED_ID FR_ID /* Component ID=0 schedules pipeline */

/*
 * Components used in static pipeline
 */

static struct sof_ipc_comp_fileread fileread_p0[] = {
	SPIPE_FILEREAD(SPIPE_COMP(FR_ID, SOF_COMP_FILEREAD,
	sof_ipc_comp_fileread), "src_in.txt"),
};

static struct sof_ipc_comp_filewrite filewrite_p0[] = {
	SPIPE_FILEWRITE(SPIPE_COMP(FW_ID, SOF_COMP_FILEWRITE,
	sof_ipc_comp_filewrite), "src_out.txt"),
};

static struct sof_ipc_comp_volume volume_p0[] = {
	SPIPE_VOL(SPIPE_COMP(VOLUME_ID, SOF_COMP_VOLUME, sof_ipc_comp_volume),
		0, 0xffffffff)
};

static struct scomps scomps_p0[] = {
	SCOMP(fileread_p0),
	SCOMP(volume_p0),
	SCOMP(filewrite_p0),
};

/*
 * Buffers used in static pipeline 2.
 */
static struct sof_ipc_buffer buffers_p0[] = {
	SPIPE_BUFFER(3, INT_PERIOD_SIZE * 2), /* B(3) */
	SPIPE_BUFFER(4, INT_PERIOD_SIZE * 2), /* B(4) */
};

/* pipeline 0 component/buffer connections */
static struct sof_ipc_pipe_comp_connect connect_p0[] = {
	SPIPE_COMP_CONNECT(0, 3), /* p(0): Fileread(0) -> B1(3) */
	SPIPE_COMP_CONNECT(3, 1), /* p(0): B1(3) ->  Volume(1) */
	SPIPE_COMP_CONNECT(1, 4), /* p(0): Volume(1) -> B24) */
	SPIPE_COMP_CONNECT(4, 2), /* p(0): B2(4) -> Filewrite(2) */
};

/* the static pipelines */
static struct spipe spipe[] = {
	SPIPE(scomps_p0, buffers_p0, connect_p0),
};

/* pipelines */
struct sof_ipc_pipe_new pipeline[] = {
	SPIPE_PIPE(0, 0, PIPE_ID, SCHED_ID, TEST_BENCH_DEADLINE, TASK_PRI_LOW, \
		TEST_BENCH_PERIOD_FRAMES),
};

void usage(char *executable);

int main(int argc, char **argv)
{
	clock_t tic, toc;
	double c_realtime;
	double t_exec;
	int fs, n_in, n_out, ret;
	struct ipc_comp_dev *pcm_dev;
	struct pipeline *p;
	struct comp_dev *cd;
	int deadline;
	struct fileread_comp_data *frcd;
	struct filewrite_comp_data *fwcd;
	int nch = TEST_BENCH_NCH; /* 2 channels test */
	int bits_in, bits_out;

	/* Handle command line */
	if (argc > 2) {
		switch (argc) {
		case 5:
			bits_in = atoi(argv[1]);
			bits_out = atoi(argv[2]);
			fileread_p0[0].fn = argv[3];
			filewrite_p0[0].fn = argv[4];
			break;
		default:
			usage(argv[0]);
			return EXIT_SUCCESS;
		}
	} else {
		usage(argv[0]);
		return EXIT_SUCCESS;
	}

	/* Init components */
	sys_comp_init();
	sys_comp_fileread_init();
	sys_comp_filewrite_init();
	sys_comp_volume_init();

	/* Set SRC sink rate into IPC, src_new() will copy it */
	fs = 48000;
	deadline = pipeline[0].deadline;
	pipeline[0].frames_per_sched = (int) (0.9999 + fs * deadline / 1e6);

	/* Set fileread config */
	switch (bits_in) {
	case 16:
		fileread_p0->config.frame_fmt = SOF_IPC_FRAME_S16_LE;
		break;
	case 24:
		fileread_p0->config.frame_fmt = SOF_IPC_FRAME_S24_4LE;
		break;
	default:
		fileread_p0->config.frame_fmt = SOF_IPC_FRAME_S32_LE;
		break;
	}

	/* Set filewrite config */
	switch (bits_out) {
	case 16:
		filewrite_p0->config.frame_fmt = SOF_IPC_FRAME_S16_LE;
		break;
	case 24:
		filewrite_p0->config.frame_fmt = SOF_IPC_FRAME_S24_4LE;
		break;
	default:
		filewrite_p0->config.frame_fmt = SOF_IPC_FRAME_S32_LE;
		break;
	}


	/* Construct pipeline, pass parameters, start */
	if (test_bench_pipeline_setup(&reef, pipeline, spipe) < 0)
		exit(EXIT_FAILURE);

	/* Params is set to source fs1 */
	if (test_bench_pipeline_params_start(reef.ipc, fs, nch, deadline, SCHED_ID, PIPE_ID) < 0)
		exit(EXIT_FAILURE);

	/* Get pointers to fileread and filewrite */
	pcm_dev = ipc_get_comp(reef.ipc, FW_ID);
	fwcd = comp_get_drvdata(pcm_dev->cd);
	pcm_dev = ipc_get_comp(reef.ipc, FR_ID);
	frcd = comp_get_drvdata(pcm_dev->cd);

	/*
	 * Run pipeline until EOF from fileread
	 */
	pcm_dev = ipc_get_comp(reef.ipc, SCHED_ID);
	p = pcm_dev->cd->pipeline;
	cd = pcm_dev->cd;
	test_bench_disable_trace(); /* Reduce print output */
	tic = clock();

	while (frcd->frs.reached_eof == 0)
		pipeline_schedule_copy(p, 0);

	/* Done! Next reset and free pipeline */
	toc = clock();
	test_bench_enable_trace();
	ret = pipeline_reset(p, cd);
	if (ret < 0) {
		printf("Failed pipeline_reset()\n");
		exit(EXIT_FAILURE);
	}
	pipeline_free(p);

	n_in = frcd->frs.n;
	n_out = fwcd->fws.n;
	t_exec = (double) (toc - tic) / CLOCKS_PER_SEC;
	c_realtime = (double) n_out / nch / fs / t_exec;
	printf("Read %d input samples,\n", n_in);
	printf("wrote %d output samples,\n", n_out);
	printf("test execution time was %.2f us, %.2f x realtime.\n", 1e3 * t_exec, c_realtime);

	return EXIT_SUCCESS;
}

void usage(char *executable)
{
	fprintf(stderr, "Usage: %s <bits in> <bits out> <input.txt> <output.txt>>\n", executable);
}

File: tune/volume/volume_test.m
function n_fail = volume_test()

%%
% volume_test - test FIR EQ executable objective audio quality parameters
%
%

%%
% Copyright (c) 2017, Intel Corporation
% All rights reserved.
%
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions are met:
%   * Redistributions of source code must retain the above copyright
%     notice, this list of conditions and the following disclaimer.
%   * Redistributions in binary form must reproduce the above copyright
%     notice, this list of conditions and the following disclaimer in the
%     documentation and/or other materials provided with the distribution.
%   * Neither the name of the Intel Corporation nor the
%     names of its contributors may be used to endorse or promote products
%     derived from this software without specific prior written permission.
%
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
% AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
% IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
% ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
% LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
% CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
% SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
% INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
% CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
% ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
% POSSIBILITY OF SUCH DAMAGE.
%
% Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
%

comp = 'volume';
addpath('../m');
plots = 'plots';
reports = 'reports';
mkdir_check(plots);
mkdir_check(reports);

ch = [1 2];
nch = 2;
fs_list = [48e3];

%% Generic test pass/fail criteria
t.g_db_tol = 0.1;
t.thdnf_db_max = -100;
t.dr_db_min = 140;

%% Loop all modes to test
t.bits_in = 16;
t.bits_out = 16;
t.ch = ch;
t.nch = nch;

n_test = 2;
n_fs = length(fs_list);
r.fs_list = fs_list;
r.g = zeros(n_fs, 1);
r.dr = zeros(n_fs, 1);
r.thdnf = zeros(n_fs, 1);
r.pf = zeros(n_fs, n_test);
r.n_fail = 0;

for a = 1:n_fs
        v = -ones(n_test,1); % Set pass/fail test verdict to not executed
        tn = 1;
        t.fs1 = fs_list(a);
        t.fs2 = fs_list(a);
        [v(1) r.g(a)] = g_test(t);
        [v(2) r.thdnf(a)] = thdnf_test(t);
        %[v(2) r.dr(a)] = dr_test(t);
        %v(4) = fr_test(t);

        %% Done, store pass/fail
        r.pf(a, :) = v;
        idx = find(v > 0);
        r.n_fail = r.n_fail + sum(v(idx));

end

%% Print table with test summary: Gain
fn = sprintf('%s/g_%s.txt', reports, comp);
print_val(fn, fs_list, r.g, r.pf, 'Gain (dB)');

%% Print table with test summary: THD+N vs. frequency
fn = sprintf('%s/thdnf_%s.txt', reports, comp);
print_val(fn, fs_list, r.thdnf, r.pf, 'Worst-case THD+N vs. frequency');

%% Print table with test summary: DR
%fn = sprintf('%s/dr_%s.txt', reports, comp);
%print_val(fn, fs_list, r.dr, r.pf, 'Dynamic range (dB CCIR-RMS)');

fprintf('\n');
fprintf('Number of failed tests = %d\n', r.n_fail);

end


%%
%% Test execution with help of common functions
%%

function [fail, thdnf] = thdnf_test(t)

%% Reference: AES17 6.3.2 THD+N ratio vs. frequency
test = test_defaults_volume(t);
test.thdnf_max = t.thdnf_db_max;

test.f_start = 20;
test.f_end = test.fs * 0.41667; % 20 kHz @ 48 kHz
test.fu = test.fs * 0.41667; % 20 kHz @ 48 kHz

%% Create input file
test = thdnf_test_input(test);

%% Run test
test = test_run_volume(test, t);

%% Measure
test = thdnf_test_measure(test);

%% For EQ use the -20dBFS result and ignore possible -1 dBFS fail
thdnf = max(test.thdnf_low);
if thdnf > test.thdnf_max
        fail = 1;
else
        fail = 0;
end
delete_check(test.fn_in);
delete_check(test.fn_out);

%% Print
volume_test_result_print(t, 'THD+N ratio vs. frequency', 'THDNF');

end


function [fail, g_db] = g_test(t)

%% Reference: AES17 6.2.2 Gain
test = test_defaults_volume(t);
test.fu = test.fs * 0.41667; % 20 kHz @ 48 kHz
test.g_db_tol = t.g_db_tol;

%% Create input file
test = g_test_input(test);

%% Run test
test = test_run_volume(test, t);

%% EQ gain at test frequency
channel = 1;
test.g_db_expect = 0;

%% Measure
test = g_test_measure(test);

%% Get output parameters
fail = test.fail;
g_db = test.g_db;
delete_check(test.fn_in);
delete_check(test.fn_out);

end

function [fail, dr_db] = dr_test(t)

%% Reference: AES17 6.4.1 Dynamic range
test = test_defaults_volume(t);
test.dr_db_min = t.dr_db_min;

%% Create input file
test = dr_test_input(test);

%% Run test
test = test_run_volume(test, t);

%% Measure
test = dr_test_measure(test);

%% Get output parameters
fail = test.fail;
dr_db = test.dr_db;
delete_check(test.fn_in);
delete_check(test.fn_out);

end



function [fail, pm_range_db, range_hz, fr3db_hz] = fr_test(t)

%% Reference: AES17 6.2.3 Frequency response
test = test_defaults_volume(t);

test.f_lo = 20;
test.f_hi = 0.999*min(t.fs1/2,t.fs2/2);
test.f_max = 0.999*min(t.fs1/2, t.fs2/2);

%% Create input file
test = fr_test_input(test);

%% Run test
test = test_run_volume(test, t);

%% Measure
test = volume_mask(test, t, 0.5);
test = fr_test_measure(test);

fail = test.fail;
pm_range_db = test.rp;
range_hz = [test.f_lo test.f_hi];
fr3db_hz = test.fr3db_hz;
delete_check(test.fn_in);
delete_check(test.fn_out);

%% Print
volume_test_result_print(t, 'Frequency response', 'FR', test.ph, test.fh);

end

function test = volume_mask(test, t, tol)
%% Create mask from theoretical frequency response calculated from coefficients
%  and align mask to be relative 997 Hz response
j = 1;
for channel = t.ch
        [b, ~] = volume_decode_blob(t.blob, t.eq(channel));
        h = freqz(b, 1, test.f, test.fs);
        m = 20*log10(abs(h));
        i997 = find(test.f > 997, 1, 'first')-1;
        m = m - m(i997);
        test.mask_f = test.f;
        test.mask_lo(:,j) = m-tol;
        test.mask_hi(:,j) = m+tol;
        j = j+1;
end

end

function test = test_defaults_volume(t)
test.bits_in = t.bits_in;
test.bits_out = t.bits_out;
test.nch = t.nch;
test.ch = t.ch;
test.fs = t.fs1;
test.mask_f = [];
test.mask_lo = [];
test.mask_hi = [];
test.rp_max = [];
end

function test = test_run_volume(test, t)
test.ex = './volume_test';
test.arg = { num2str(t.bits_in), num2str(t.bits_out), num2str(test.fn_in), num2str(test.fn_out)};
delete_check(test.fn_out);
test = test_run(test);
end

function volume_test_result_print(t, testverbose, testacronym, ph, fh)
tstr = sprintf('%s Volume %d Hz', testverbose, t.fs1);
if nargin > 3
        nph = length(ph);
        for i=1:nph
                title(ph(i), tstr);
        end
else
        title(tstr);
end
if nargin > 4
        nfh = length(fh);
        for i=1:nfh
                figure(fh(i));
                pfn = sprintf('plots/%s_volume_%d_%d.png', ...
                        testacronym, t.fs1, i);
                print(pfn, '-dpng');
        end
else
        pfn = sprintf('plots/%s_volume_%d.png', testacronym, t.fs1);
        print(pfn, '-dpng');
end
end


%%
%% The next are results printing functions
%%

function  print_val(fn, fs_list, val, pf, valstr)
n_fs = length(fs_list);
fh = fopen(fn,'w');
fprintf(fh,'\n');
%fprintf(fh,'%s\n', valstr);
fprintf(fh,'%8s, ', 'Fs (kHz)');
fprintf(fh,'%s', valstr);
fprintf(fh,'\n');
for a = 1:n_fs
        fprintf(fh,'%8.1f, ', fs_list(a)/1e3);
        if pf(a,1) < 0
                cstr = 'x';
        else
                if isnan(val(a))
                        cstr = '-';
                else
                        cstr = sprintf('%8.2f', val(a));
                end
        end
        fprintf(fh,'%8s', cstr);
        fprintf(fh,'\n');
end
fclose(fh);
type(fn);
end

function print_fr(fn, fs_in_list, fs_out_list, fr_db, fr_hz, pf)
n_fsi = length(fs_in_list);
n_fso = length(fs_out_list);
fh = fopen(fn,'w');
fprintf(fh,'\n');
fprintf(fh,'SRC test result: Frequency response +/- X.XX dB (YY.Y kHz) \n');
fprintf(fh,'%8s, ', 'out \ in');
for a = 1:n_fsi-1
        fprintf(fh,'%12.1f, ', fs_in_list(a)/1e3);
end
fprintf(fh,'%12.1f', fs_in_list(n_fsi)/1e3);
fprintf(fh,'\n');
for b = 1:n_fso
        fprintf(fh,'%8.1f, ', fs_out_list(b)/1e3);
        for a = 1:n_fsi
                if pf(a,b,1) < 0
                        cstr = 'x';
                else
                        cstr = sprintf('%4.2f (%4.1f)', ...
                                fr_db(a,b), fr_hz(a,b,2)/1e3);
                end
                if a < n_fsi
                        fprintf(fh,'%12s, ', cstr);
                else
                        fprintf(fh,'%12s', cstr);
                end
        end
        fprintf(fh,'\n');
end
fclose(fh);
type(fn);
end

function print_fr3db(fn, fs_in_list, fs_out_list, fr3db_hz, pf)
n_fsi = length(fs_in_list);
n_fso = length(fs_out_list);
fh = fopen(fn,'w');
fprintf(fh,'\n');
fprintf(fh,'SRC test result: Frequency response -3dB 0 - X.XX kHz\n');
fprintf(fh,'%8s, ', 'out \ in');
for a = 1:n_fsi-1
        fprintf(fh,'%8.1f, ', fs_in_list(a)/1e3);
end
fprintf(fh,'%8.1f', fs_in_list(n_fsi)/1e3);
fprintf(fh,'\n');
for b = 1:n_fso
        fprintf(fh,'%8.1f, ', fs_out_list(b)/1e3);
        for a = 1:n_fsi
                if pf(a,b,1) < 0
                        cstr = 'x';
                else
                        cstr = sprintf('%4.1f', fr3db_hz(a,b)/1e3);
                end
                if a < n_fsi
                        fprintf(fh,'%8s, ', cstr);
                else
                        fprintf(fh,'%8s', cstr);
                end
        end
        fprintf(fh,'\n');
end
fclose(fh);
type(fn);
end


function print_pf(fn, fs_in_list, fs_out_list, pf)
n_fsi = length(fs_in_list);
n_fso = length(fs_out_list);
fh = fopen(fn,'w');
fprintf(fh,'\n');
fprintf(fh,'SRC test result: Fails\n');
fprintf(fh,'%8s, ', 'out \ in');
for a = 1:n_fsi-1
        fprintf(fh,'%14.1f, ', fs_in_list(a)/1e3);
end
fprintf(fh,'%14.1f', fs_in_list(n_fsi)/1e3);
fprintf(fh,'\n');
spf = size(pf);
npf = spf(3);
for b = 1:n_fso
        fprintf(fh,'%8.1f, ', fs_out_list(b)/1e3);
        for a = 1:n_fsi
                if pf(a,b,1) < 0
                        cstr = 'x';
                else
                        cstr = sprintf('%d', pf(a,b,1));
                        for n=2:npf
                                if pf(a,b,n) < 0
                                        cstr = sprintf('%s/x', cstr);
                                else
                                        cstr = sprintf('%s/%d', cstr,pf(a,b,n));
                                end
                        end
                end
                if a < n_fsi
                        fprintf(fh,'%14s, ', cstr);
                else
                        fprintf(fh,'%14s', cstr);
                end
        end
        fprintf(fh,'\n');
end
fclose(fh);
type(fn);
end
